RBF move 2/3: extract RBF logic into policy/rbf

glozow opened the pull request:
This PR does not change behavior. It extracts the BIP125 logic into helper functions (and puts them in the policy/rbf* files). This enables three things - I think each one individually is pretty good:

- Implementation of package RBF (see #22290). I want it to be as close to BIP125 as possible so that it doesn't become a distinct fee-bumping mechanism. Doing these move-only commits first means the diff is mostly mechanical to review, and I just need to create a function that mirrors the single transaction validation.
- We will be able to isolate and test our RBF logic alone. Recently, there have been some discussions on discrepancies between our code and BIP125, as well as proposals for improving it. Generally, I think making this code more modular and de-bloating validation.cpp is probably a good idea.
- Witness Replacement (replacing same-txid-different-wtxid when the witness is significantly smaller and therefore higher feerate) in a BIP125-similar way. Hopefully it can just be implemented with calls to the rbf functions (i.e. `PaysForRBF`) and an edit to the relevant mempool entries.

glozow committed:
MOVEONLY: check for disjoint conflicts and ancestors to policy/rbf

This checks that a transaction isn't trying to replace something it
supposedly depends on.

glozow committed:
MOVEONLY: BIP125 Rule 2 to policy/rbf

0xB10C commented:
Concept ACK

TIL Witness Replacement

glozow commented:
A new circular dependency `policy/rbf -> txmempool -> validation -> policy/rbf` is introduced here, so I've added it to `EXPECTED_CIRCULAR_DEPENDENCIES` to suppress the issue. I understand this is not ideal. The main problem, I think, is the txmempool -> validation part (which already exists), so I've taken a crack at it in #22677.

fanquake commented:
Concept ACK

mjdietzx commented on src/policy/rbf.cpp:
Note: `nConflictingSize` is an unused variable

glozow replied:
Removed - thanks!

mjdietzx reviewed (commented):
Concept ACK. Very well done, I think this is a nice improvement

glozow commented on src/policy/rbf.cpp:
```diff
@@ -47,6 +47,29 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
     return SignalsOptInRBF(tx) ? RBFTransactionState::REPLACEABLE_BIP125 : RBFTransactionState::UNKNOWN;
 }
 
+bool IsRBFOptOut(const CTransaction& txConflicting)
```
Hm, I think everything in util/rbf should be moved to policy/rbf actually :thinking: There's no reason to have two places for RBF (I think) and organization-wise, all RBF logic and code is policy.

darosior replied:
Not sure, it would make all users of policy/rbf depend on txmempool. Eg not sure the standalone binaries `wallet-util` and friends depend on it although they use util/rbf.

glozow replied:
Ah you're right @darosior

glozow commented on src/policy/rbf.h:
```diff
@@ -7,6 +7,10 @@
 
+/** Maximum number of transactions that can be replaced by BIP125 RBF (Rule #5). This includes all
+ * mempool conflicts and their descendants. */
+static constexpr uint32_t MAX_BIP125_REPLACEMENT_CANDIDATES{100};
```
It wouldn't get rid of the circular dependency unfortunately, because it's caused by validation.cpp including policy/rbf.h

glozow commented on src/policy/rbf.cpp:
```diff
@@ -42,3 +46,38 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
+        }
+
+        nConflictingCount += mi->GetCountWithDescendants();
+        // This potentially overestimates the number of actual descendants
```
Yes, that's exactly right.

glozow committed:
MOVEONLY: check that fees > direct conflicts to policy/rbf

glozow committed:
MOVEONLY: fee checks (Rules 3 and 4) to policy/rbf

glozow commented on src/policy/rbf.cpp:
Oopsie good catch! did a few different attempts at moving BIP125, this is leftover from a previous one :sweat_smile:

mjdietzx commented:
crACK 6a0a8ef791400d78fb4192a81a392131be78270b

theStack commented:
Concept ACK

mjdietzx commented:
Tested ACK 6a0a8ef791400d78fb4192a81a392131be78270b

theStack commented on src/policy/rbf.cpp:
```diff
@@ -43,3 +47,31 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
 }
 
+bool GetEntriesForRBF(const CTransaction& tx, CTxMemPool& m_pool,
+                      const CTxMemPool::setEntries setIterConflicting,
```
I guess for performance reasons, we should pass `setIterConflicting` by reference, rather than by value?
```suggestion
                      const CTxMemPool::setEntries& setIterConflicting,
```

glozow replied:
Good point, done!

theStack commented on src/policy/rbf.cpp:
```diff
@@ -2,9 +2,13 @@
 #include <policy/rbf.h>
 #include <util/rbf.h>
 
+#include <string>
```
This header doesn't seem to be directly needed in this compilation unit (strings _are_ used as return values/parameters for `uint256.ToString()` and `strprintf`, but it's the job of the headers offering those functions to already also include `<string>`).
```suggestion
```

glozow replied:
Done

theStack reviewed (commented):
Reviewed up to 474258f4296878a3740c8e19c9358ba4134cd48d (5/9 commits) so far, LGTM, nice cleanups. To be extra-cautious, I also compiled and ran unit tests + the functional test feature_rbf.py on every commit.

Left two minor nits. Planning to review the remaining 4 commits tomorrow.

theStack commented on src/policy/rbf.cpp:
```diff
@@ -75,3 +75,40 @@ bool GetEntriesForRBF(const CTransaction& tx, CTxMemPool& m_pool,
     return true;
 }
 
+bool HasNoNewUnconfirmed(const CTransaction& tx, CTxMemPool& m_pool,
+                         const CTxMemPool::setEntries setIterConflicting, TxValidationState& state)
```
```suggestion
bool HasNoNewUnconfirmed(const CTransaction& tx, const CTxMemPool& m_pool,
                         const CTxMemPool::setEntries& setIterConflicting, TxValidationState& state)
```

glozow replied:
Done

theStack commented on src/policy/rbf.cpp:
```diff
@@ -112,3 +112,20 @@ bool HasNoNewUnconfirmed(const CTransaction& tx, CTxMemPool& m_pool,
     return true;
 }
+
+bool SpendsAndConflictsDisjoint(CTxMemPool::setEntries& setAncestors, std::set<uint256> setConflicts,
```
```suggestion
bool SpendsAndConflictsDisjoint(const CTxMemPool::setEntries& setAncestors, const std::set<uint256>& setConflicts,
```

glozow replied:
Good point, done

theStack commented on src/policy/rbf.h:
```diff
@@ -51,4 +51,12 @@ bool GetEntriesForRBF(const CTransaction& tx, CTxMemPool& m_pool,
+/** Check the intersection between original mempool transactions (candidates for being replaced) and
+ * the ancestors of replacement transactions.
+ * @param[in]   hash    Transaction ID, included in the error message if violation occurs.
+ * returns false if the intersection is empty, true if otherwise.
```
```suggestion
 * returns true if the intersection is empty, false if otherwise.
```

glozow replied:
Fixed

theStack commented on src/validation.cpp:
```diff
@@ -828,7 +804,6 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
             return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "insufficient fee",
                     strprintf("rejecting replacement %s, not enough additional fees to relay; %s < %s",
                         hash.ToString(),
-                        FormatMoney(nDeltaFees),
```
This line removal must be unintentional (three placeholders, but only two arguments)? Quite a pity that the compiler doesn't warn us in this case. At least the tinyformat library seems to be nice enough to notice us at runtime :)
```
$ ./test/functional/feature_rbf.py
.....
.....
2021-08-14T18:32:49.823000Z TestFramework (INFO): Running test replacement relay fee...
2021-08-14T18:32:49.834000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):                                                                                                                                                             File "/home/honey/bitcoin_prrev/test/functional/test_framework/util.py", line 132, in try_rpc
    fun(*args, **kwds)
  File "/home/honey/bitcoin_prrev/test/functional/test_framework/coverage.py", line 49, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/honey/bitcoin_prrev/test/functional/test_framework/authproxy.py", line 146, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: tinyformat: Too many conversion specifiers in format string (-1)
```

glozow replied:
woah good catch O.o

glozow replied:
Fixed

theStack reviewed (commented):
Left some more suggestions considering const-correctness and passing by reference rather than by value. On the second-to-last commit a30b0270d0129f16573827d45250aa42912803d3 the test feature_rbf.py currently fails, see comment below. In the last commit this is fixed, as the deleted strprintf() argument line is re-introduced in the helper function.

glozow committed:
scripted-diff: rename variables in policy/rbf

"Fee Delta" is already a term used for prioritizing transactions:
modified = base fees + delta

Here, delta also means the difference between original and modified replacement fees:
nDeltaFees = (original_base + original_delta) - (replacement_base + replacement_delta)

This is insanely confusing. Also, since mempool is no longer a member of a
class (MemPoolAccept.m_pool), the "m" prefix is unnecessary. The rest are
clarity/style-focused changes to already-touched lines.

-BEGIN VERIFY SCRIPT-

ren() { sed -i "s/\<$1\>/$2/g" src/policy/rbf* ; }

ren nDeltaFees additional_fees
ren m_pool pool

ren nSize replacement_vsize
ren nModifiedFees replacement_fees
ren nConflictingFees original_fees
ren oldFeeRate original_feerate
ren newFeeRate replacement_feerate

ren setAncestors ancestors
ren setIterConflicting iters_conflicting
ren setConflictsParents parents_of_conflicts
ren setConflicts direct_conflicts
ren allConflicting all_conflicts

sed -i "s/ hash\b/ txid/g" src/policy/rbf*
-END VERIFY SCRIPT-

glozow commented:
> To be extra-cautious, I also compiled and ran unit tests + the functional test feature_rbf.py on every commit.
> Left some more suggestions considering const-correctness and passing by reference rather than by value. On the second-to-last commit [a30b027](https://github.com/bitcoin/bitcoin/commit/a30b0270d0129f16573827d45250aa42912803d3) the test feature_rbf.py currently fails, see comment below. In the last commit this is fixed, as the deleted strprintf() argument line is re-introduced in the helper function.

Thanks @theStack and good catch! Fixed. Good idea for reviewers to run `git rebase --exec "make check; test/functional/feature_rbf.py" -i HEAD~11` in general

Also added a small scripted diff style cleanup, should be simple to review

laanwj commented on test/lint/lint-circular-dependencies.sh:
```diff
@@ -15,6 +15,7 @@ EXPECTED_CIRCULAR_DEPENDENCIES=(
     "index/base -> validation -> index/blockfilterindex -> index/base"
     "index/coinstatsindex -> node/coinstats -> index/coinstatsindex"
     "policy/fees -> txmempool -> policy/fees"
+    "policy/rbf -> txmempool -> validation -> policy/rbf"
```
Concept ACK, but I think we should avoid adding a circular dependency here.

glozow replied:
Yep! I've rebased #22677 on top of this PR so it's more clear that it would fix this problem.

glozow replied:
Okie done

theStack reviewed (approved):
Code-review ACK 192193c7c420ebbbac6bb0a60778e6a352b3d67c

Verified via `git range-diff 6a0a8ef7...192193c7` that since my last review (https://github.com/bitcoin/bitcoin/pull/22675#pullrequestreview-729819432 and https://github.com/bitcoin/bitcoin/pull/22675#pullrequestreview-730123757) all suggestions have been tackled and that the newly added commits 7d6571ffed165d681c8b016b965aa94f41cfadb0 (renames) and 192193c7c420ebbbac6bb0a60778e6a352b3d67c (style cleanups) are okay. Kudos for the new variable names, I think the code is much more comprehensible now 👌 

> Good idea for reviewers to run `git rebase --exec "make check; test/functional/feature_rbf.py" -i HEAD~11` in general

Oh nice, I didn't know about `git rebase --exec` before (I always did review/compile/run test manually hopping from commit to commit), that is a huge time-saver! 🚀

ariard reviewed (commented):
Concept ACK, I like the direction to drain `PreChecks` in isolated logical components and adding clearer documentation in the process.

I think the changes proposed by this PR are sizeable enough to be splitted off in different PRs.  Especially about the replace-by-fee check rules, beyond the better-feerate-than-directly-replaced transaction present in the code but not in the spec, iirc they're still minor discrepancies worthy to be documented.

One PRs-split-offs could be :
* `IsRBFOptOut()` : the replacement signaling, introducing `src/policy/rbf.h`
* `SpendsAndConflictsDisjoint` : the replacement candidate sanitization
* `PaysMoreThanConflicts`/... : the replacement check rules

darosior reviewed (commented):
Concept ACK nice cleanup.

I was not sure about the approach at first, since we don't currently *need* `txmempool` in `policy/rbf` for `IsRBFOptIn` (#22665 removes it). But i agree it's cleaner to rather have `txmempool` not depend on `validation` as you do in your followup.
That said, why not proceed the other way around to avoid introducing the circular dependency at all?

glozow commented:
> That said, why not proceed the other way around to avoid introducing the circular dependency at all?

True, I just did it this way because this PR is a bit further along (other one is still looking for approach feedback), and I have some branches depending on this refactor.

jnewbery commented on src/policy/rbf.cpp:
```diff
@@ -42,3 +46,133 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
+    for (unsigned int j = 0; j < tx.vin.size(); j++)
+    {
+        // We don't want to accept replacements that require low feerate junk to be mined first.
+        // Ideally we'd keep track of the ancestor feerates and make the decision based on that, but
+        // for now requiring all new inputs to be confirmed works.
```
Maybe explicitly mention that this is BIP125 rule 2.

jnewbery commented on src/util/rbf.cpp:
```diff
@@ -15,3 +15,26 @@ bool SignalsOptInRBF(const CTransaction &tx)
+
+bool IsRBFOptOut(const CTransaction& txConflicting)
+{
+    // Allow opt-out of transaction replacement by setting
+    // nSequence > MAX_BIP125_RBF_SEQUENCE (SEQUENCE_FINAL-2) on all inputs.
```
Maybe explicitly comment that this is BIP125 rule 1.

jnewbery commented on src/policy/rbf.cpp:
```diff
@@ -42,3 +46,133 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
+        // This potentially overestimates the number of actual descendants but we just want to be
+        // conservative to avoid doing too much work.
+        if (nConflictingCount > MAX_BIP125_REPLACEMENT_CANDIDATES) {
+            return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too many potential replacements",
```
Maybe explicitly comment that this is BIP125 rule 5.

jnewbery commented on src/util/rbf.h:
```diff
@@ -15,4 +15,7 @@ static const uint32_t MAX_BIP125_RBF_SEQUENCE = 0xfffffffd;
 
+/** Determine whether a mempool transaction is opting out of RBF. Mempool entries do not inherit
+ * signaling from their parents in this implementation. */
+bool IsRBFOptOut(const CTransaction& txConflicting);
```
Since you're changing the parameter name already, consider:

```suggestion
bool IsRBFOptOut(const CTransaction& tx);
```

glozow replied:
Deleted this, ended up just calling `SignalsOptInRBF` instead

jnewbery reviewed (commented):
Looks good!

The logs for some of the commits could use a little more detail, e.g. e535422365 _[validation] quit earlier and separate loops_ could have a better summary (Quit what earlier? Separate which loops? At the very least this should mention ATMP or RBF). The same is true for some of the other commit logs as well.

glozow commented on src/policy/rbf.h:
```diff
@@ -31,4 +36,51 @@ enum class RBFTransactionState {
+ * @param[in]   iters_conflicting   The set of iterators to mempool entries.
+ * @param[out]  state               Used to return errors, if any.
+ * @param[out]  all_conflicts       Populated with all the mempool entries that would be replaced,
+ *                                  which includes descendants of iters_conflicting. Not cleared at
+ *                                  the start; any existing mempool entries will remain in the set.
```
![image](https://user-images.githubusercontent.com/25183001/130248786-e8c84e2f-4932-4605-b372-fb5c9edb4976.png)

glozow replied:
https://github.com/bitcoin/bitcoin/pull/22290/commits/6c04bf5a60e2af86c326c4e65aba777d8bf768bb#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R970-R972

glozow commented on src/util/rbf.cpp:
(Doesn't apply anymore because I deleted the function)

glozow commented on src/validation.cpp:
```diff
@@ -784,23 +761,8 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
-    // Check if it's economically rational to mine this transaction rather
-    // than the ones it replaces.
-    nConflictingFees = 0;
-    nConflictingSize = 0;
```
Done

glozow commented on src/policy/rbf.h:
```diff
@@ -6,6 +6,11 @@
 #define BITCOIN_POLICY_RBF_H
 
 #include <txmempool.h>
+#include <consensus/validation.h>
```
Done

glozow replied:
Actually decided to remove this dependency because I realized it's not the right place to be deciding what `TxValidationResult` to return (should be in validation.cpp)
But yes, sorted includes :salute:

glozow commented on src/validation.cpp:
```diff
@@ -809,106 +771,21 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-        }
+        // Calculate all conflicting entries and enforce Rules 2 and 5.
+        if (!GetEntriesForRBF(tx, m_pool, setIterConflicting, state, allConflicting)) return false;
+        if (!HasNoNewUnconfirmed(tx, m_pool, setIterConflicting, state)) return false;
```
Done, good point

glozow commented on src/policy/rbf.cpp:
```diff
@@ -2,14 +2,18 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <consensus/validation.h>
+#include <logging.h>
```
Switched to tinyformat

glozow commented on src/policy/rbf.cpp:
```diff
@@ -42,3 +46,133 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
+bool PaysForRBF(CAmount original_fees, CAmount replacement_fees, size_t replacement_vsize,
+                TxValidationState& state, const uint256& hash)
+{
+    // The replacement must pay greater fees than the transactions it replaces - if we did the
```
Marked all of them with their corresponding BIP125 rule numbers :+1:

glozow commented on src/policy/rbf.cpp:
```diff
@@ -42,3 +46,133 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
+    for (CTxMemPool::txiter ancestorIt : ancestors) {
+        const uint256 &hashAncestor = ancestorIt->GetTx().GetHash();
+        if (direct_conflicts.count(hashAncestor)) {
+            return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-txns-spends-conflicting-tx",
```
good eyes. added a comment

glozow commented on src/policy/rbf.h:
```diff
@@ -31,4 +36,51 @@ enum class RBFTransactionState {
+ * @param[in]   ancestors           Set of mempool entries corresponding to ancestors of the
+ *                                  replacement transactions.
+ * @param[in]   direct_conflicts    Set of mempool entries corresponding to the mempool conflicts
+ *                                  (candidates to be replaced).
```
Done and fixed the commit jumping

glozow commented on src/policy/rbf.h:
```diff
@@ -31,4 +36,51 @@ enum class RBFTransactionState {
+ *                                  the start; any existing mempool entries will remain in the set.
+ * @returns false if Rule 5 is broken.
+ */
+bool GetEntriesForRBF(const CTransaction& tx, CTxMemPool& pool,
```
I named it `GetEntriesForConflicts()`

glozow commented on src/policy/rbf.h:
```diff
@@ -31,4 +36,51 @@ enum class RBFTransactionState {
+ * @param[in]   hash                Transaction ID, included in the error message if violation occurs.
+ * returns true if the two sets are disjoint (i.e. intersection is empty), false if otherwise.
+ */
+bool SpendsAndConflictsDisjoint(const CTxMemPool::setEntries& ancestors,
```
I didn't end up renaming `hash` to `txid` because I want to be able to pass in either package id or txid in the future :thinking: let me know if it's too weird though, and I can change it

jnewbery replied:
I'd personally prefer fewer surprise tools that help us later, since it's easy enough to update comments/rename parameters when you actually need that functionality. No big deal either way though!

glozow replied:
renamed to txid

glozow commented on src/policy/rbf.h:
```diff
@@ -31,4 +36,51 @@ enum class RBFTransactionState {
+                      const CTxMemPool::setEntries& iters_conflicting, TxValidationState& state,
+                      CTxMemPool::setEntries& all_conflicts) EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
+
+/** Enforce BIP125 Rule 2: a replacement transaction must not add any new unconfirmed inputs. */
```
Makes sense, quoted BIP125

jnewbery reviewed (commented):
In commit 76828aa9c1 _MOVEONLY: fee checks (Rules 3 and 4) to policy/rbf_, you update the doxygen comment for `HasNoNewUnconfirmed()`. That change should have been squashed into commit 526b490e81 _MOVEONLY: BIP125 Rule 2 to policy/rbf_.

jnewbery commented on src/policy/rbf.cpp:
```diff
@@ -42,3 +46,133 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
+{
+    AssertLockHeld(pool.cs);
+    const uint256 hash = tx.GetHash();
+    uint64_t nConflictingCount = 0;
```
(not addressed)

glozow replied:
(will fix in a later PR)

jnewbery commented on src/policy/rbf.h:
```diff
@@ -31,4 +36,51 @@ enum class RBFTransactionState {
+                         const CTxMemPool::setEntries& iters_conflicting,
+                         TxValidationState& state) EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
+
+/** Check the intersection between two sets of mempool entries to make sure they are disjoint.
```
(not addressed - this comment still refers to "two sets of mempool entries")

glozow replied:
now calling it "two sets of transactions (a set of mempool entries and a set of txids)"

glozow commented on src/policy/rbf.cpp:
```diff
@@ -3,13 +3,17 @@
+#include <policy/settings.h>
+#include <util/moneystr.h>
 #include <util/rbf.h>
+#include <tinyformat.h>
```
Fixt

glozow commented on src/validation.cpp:
```diff
@@ -784,130 +770,45 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-            }
+        // Calculate all conflicting entries and enforce Rule #5.
+        if (!GetEntriesForConflicts(tx, m_pool, setIterConflicting, allConflicting, err_string)) {
+            return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, err_string);
```
:pleading_face: Added it back

glozow commented:
Addressed @jnewbery comments

theStack commented:
I tend to agree with @ariard that splitting this PR into different ones would make a lot of sense here:

> I think the changes proposed by this PR are sizeable enough to be splitted off in different PRs. Especially about the replace-by-fee check rules, beyond the better-feerate-than-directly-replaced transaction present in the code but not in the spec, iirc they're still minor discrepancies worthy to be documented.
> 
> One PRs-split-offs could be :
> 
> * `IsRBFOptOut()` : the replacement signaling, introducing `src/policy/rbf.h`
> * `SpendsAndConflictsDisjoint` : the replacement candidate sanitization
> * `PaysMoreThanConflicts`/... : the replacement check rules

Maybe it's my lousy git skills, but I just tried to re-review the changes since my last ACK via `git range-diff` and the results are rather long (> 800 LOC) and confusing (e.g. some commits have been completely removed), to a point that a complete new review is probably quicker.

jnewbery commented:
Code review ACK a33fdd0b981965982754b39586eedb7ae456ba57

I'm very happy to review any new PRs if you split this up.

glozow commented:
I've split off the first few commits into #22796.

glozow committed:
whitespace fixups after move and scripted-diff

glozow commented:
Rebased and split off the documentation changes. Ready for review.
Please note that this is move-only, and I'm happy to add any doc suggestions to #22855.

glozow committed:
Make GetEntriesForConflicts return std::optional

Avoids reusing err_string.

glozow commented:
Changed the functions to return `std::optional<std::string>` instead of `bool` in response to https://github.com/bitcoin/bitcoin/pull/22796#discussion_r700944084

mjdietzx commented:
ACK 32748da0f47f7aa9fba78dfb29aa426b14f15624

I tested this again by running some additional test coverage around BIP125 (this PR #22867) on this branch. All good ✅

theStack reviewed (approved):
Code-review ACK 32748da0f47f7aa9fba78dfb29aa426b14f15624 📐 

Built and ran unit tests and functional test `./test/functional/feature_rbf.py` at each commit, verified by review that there is no change in behaviour. Here are some naming/documentation improvement nits:
* to increase clarity, you could explicitely refer to BIP125 when mentioning rules in the comments in `MemPoolAccept::PreChecks()`, e.g. `s/Enforce Rule #2/Enforce BIP125 Rule #2/`
* though I totally understand the motivation behind the new interface, returning `std::optional<std::string>` seems a bit confusing for functions that previously returned `bool`. The function name is like a predicate, and the predicate (e.g. `PaysForRBF`) is now unexpectedly true if _nothing_ (std::nullopt) is returned, leading to an inverted logic for the reader of the if-clause where it is called. Maybe prefix the function names with something like `Verify` or `Check`?

MarcoFalke commented on src/policy/rbf.cpp:
```diff
@@ -161,3 +161,29 @@ std::optional<std::string> PaysMoreThanConflicts(const CTxMemPool::setEntries& s
+    // Finally in addition to paying more fees than the conflicts the
+    // new transaction must pay for its own bandwidth.
+    CAmount nDeltaFees = nModifiedFees - nConflictingFees;
+    if (nDeltaFees < ::incrementalRelayFee.GetFee(nSize))
```
ac761f0a23c9c469fa00885edf3d5c9ae7c6a2b3: Would be nice to not use a global here. Testability is one of the motivations in OP, but using globals makes testing actually harder.

glozow replied:
Yes definitely agree - I can do this in the followup PR.

glozow replied:
Done in https://github.com/bitcoin/bitcoin/pull/22855/commits/7ccaac6c5332b8a93c9b5c7103f846854ee595d5

maflcko reviewed (approved):
review ACK 32748da0f47f7aa9fba78dfb29aa426b14f15624 🦇

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 32748da0f47f7aa9fba78dfb29aa426b14f15624 🦇
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjoaQwAggcrCYBYlCDfIy4FYpWzVg9dKVuvdwpkRXNo146y0z4w2cNoJa+08akW
UfBbdsu2YlwbPKlgODU6LARsjq5PE8nj+vkj1S3Z9MnvhoXh/UwLlRci2bDDshn1
YyiTdtiFfN1UYFQ4ShhunipzGK6ZNALWZUDFXCIWbtVQUYS0shmWls/Dchah94lf
2hmK9vGXLbtpAuR/9zxYguUL9ixcsBtALCI3GG2DspZIHON3aJBZDP59HJXXINo+
aQoFNw77QfQKa+GMKLvCW8/5eQDVue5xAJyckB+EyJjXE2ztcskn2riA+VP+I0Zu
PfVcqgkfyy5Ci/8BUjvtWAIYGa+WUeJjbtd/DVw80E9CpqILj3TPvrpTeA2q3bHC
PAJQMrmZbEGQgMir9Zj/c9Bh54+9MnEhKwo5JqktBI110ohJsIkFwuT75hXrKRjn
53jrMg7LR9kzA2Hmg625hgxJTNthv1QMZrerkhRjA1ad/s1bE/hGnN2Lk5yYy/8k
FATeV8qZ
=rsJ0
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `23997a2a7eb742c9c9782006676c24ca77f830db8c1bb6251e751a60a1f59bc5  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010823997a2a7eb742c9c9782006676c24ca77f830db8c1bb6251e751a60a1f59bc5f010bca5765aacfaccab3caa80bb8bdd232608fff01011bd6f84ef613c1cdfbc7c9948e2864a08f1046135dc25f0089eb0e8a2a01bc1c40083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01063e66da39a3db96566b8c69cec58666608f120fabc39524e791d0356d520afbc71120d3399f20bce054fc288917c04757710eb08f1046135dc24f008f6bdc60a8910c8310083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010702484a797a3f9fe9da6d14583a61c1508f1046135dc24f008afdf00c0356e0e330083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010896da09b3403101142a040f50ebf567b08f1046135dc24f008ad5bf53f120f39050083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>


The scripted diff might also modify out-of-tree files. You can fix that by passing the glob through `git grep -l`:

```
sed ... $(git grep -l "$1" ./src/policy/rbf*)
```

Also in the last commit, if you fixup the file so much, might as well go the extra step and fixup the whole file's whitespace:

```diff
diff --git a/src/policy/rbf.cpp b/src/policy/rbf.cpp
index 15527afb8a..d524b95b89 100644
--- a/src/policy/rbf.cpp
+++ b/src/policy/rbf.cpp
@@ -48,9 +48,9 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
 }
 
 std::optional<std::string> GetEntriesForConflicts(const CTransaction& tx,
-                            CTxMemPool& pool,
-                            const CTxMemPool::setEntries& iters_conflicting,
-                            CTxMemPool::setEntries& all_conflicts)
+                                                  CTxMemPool& pool,
+                                                  const CTxMemPool::setEntries& iters_conflicting,
+                                                  CTxMemPool::setEntries& all_conflicts)
 {
     AssertLockHeld(pool.cs);
     const uint256 txid = tx.GetHash();
@@ -81,7 +81,7 @@ std::optional<std::string> HasNoNewUnconfirmed(const CTransaction& tx,
     AssertLockHeld(pool.cs);
     std::set<uint256> parents_of_conflicts;
     for (const auto& mi : iters_conflicting) {
-        for (const CTxIn &txin : mi->GetTx().vin) {
+        for (const CTxIn& txin : mi->GetTx().vin) {
             parents_of_conflicts.insert(txin.prevout.hash);
         }
     }
@@ -111,7 +111,7 @@ std::optional<std::string> EntriesAndTxidsDisjoint(const CTxMemPool::setEntries&
                                                    const uint256& txid)
 {
     for (CTxMemPool::txiter ancestorIt : ancestors) {
-        const uint256 &hashAncestor = ancestorIt->GetTx().GetHash();
+        const uint256& hashAncestor = ancestorIt->GetTx().GetHash();
         if (direct_conflicts.count(hashAncestor)) {
             return strprintf("%s spends conflicting transaction %s",
                              txid.ToString(),
diff --git a/src/policy/rbf.h b/src/policy/rbf.h
index 56468a09b2..3f37bb09bf 100644
--- a/src/policy/rbf.h
+++ b/src/policy/rbf.h
@@ -48,14 +48,14 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx);
 std::optional<std::string> GetEntriesForConflicts(const CTransaction& tx, CTxMemPool& pool,
                                                   const CTxMemPool::setEntries& iters_conflicting,
                                                   CTxMemPool::setEntries& all_conflicts)
-                                                  EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
+    EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
 
 /** BIP125 Rule #2: "The replacement transaction may only include an unconfirmed input if that input
  * was included in one of the original transactions."
  * @returns error message if Rule #2 is broken, otherwise std::nullopt. */
 std::optional<std::string> HasNoNewUnconfirmed(const CTransaction& tx, const CTxMemPool& pool,
                                                const CTxMemPool::setEntries& iters_conflicting)
-                                               EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
+    EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
 
 /** Check the intersection between two sets of transactions (a set of mempool entries and a set of
  * txids) to make sure they are disjoint.
diff --git a/src/util/rbf.h b/src/util/rbf.h
index 6d44a2cb83..db9549e3ed 100644
--- a/src/util/rbf.h
+++ b/src/util/rbf.h
@@ -18,6 +18,6 @@ static const uint32_t MAX_BIP125_RBF_SEQUENCE = 0xfffffffd;
 * SEQUENCE_FINAL-1 is picked to still allow use of nLockTime by non-replaceable transactions. All
 * inputs rather than just one is for the sake of multi-party protocols, where we don't want a single
 * party to be able to disable replacement. */
-bool SignalsOptInRBF(const CTransaction &tx);
+bool SignalsOptInRBF(const CTransaction& tx);
 
 #endif // BITCOIN_UTIL_RBF_H

ariard commented on src/policy/rbf.h:
```diff
@@ -35,19 +35,61 @@ enum class RBFTransactionState {
+/** Enforce BIP125 Rule #3 "The replacement transaction pays an absolute fee of at least the sum
+ * paid by the original transactions." Enforce BIP125 Rule #4 "The replacement transaction must also
+ * pay for its own bandwidth at or above the rate set by the node's minimum relay fee setting."
+ * @param[in]   original_fees       Total modified fees of original transaction(s).
```
Note, as previously mentioned, I think this is where we do have additional divergences w.r.t to bip125.

Rule 3 says "The replacement transaction pays an absolute fee of at least the sum paid by the original transactions".

I think original transactions as to be interpreted as the set of directly conflicted transactions, i.e sharing inputs with the replacement transaction. At least Rule 5 explicitly dissociates "original transactions to be replaced and their descendant transactions".

However our RBF implementation computes the `original_fees` from iterating on `allConflicting` which is built from `CalculateDescendants`. So to-be-replaced descendants fees are also added to the sum paid checked.

This means our implementation is more conservative than BIP125, as the replacement threshold is higher. I think this is preferable, at least without deeper thinking.

AFAIK, this is mental model shared by most developers when they reason about our RBF logic (e.g in pinning discussions). Though it might create surprises if a user submits replacement transactions to Core and other full-node like bcoin, btcd (I don't know their behaviors), they might be refused by Core but accepted in other implems.

If I'm correct, good to make this point more known.

darosior replied:
You are correct with regard to the implementation. However BIP125 says:
> The replacement transaction pays an absolute fee of at least the sum paid by the original transactions

Which is plural, so i think the implementation is in line with this rule.
Of course, i agree that it would be a nice place to look for bugs in re-implementations :)

glozow replied:
At least wrt fees, it wouldn't make much sense to not consider the descendants of direct conflicts. You can very easily end up replacing higher-fee packages with lower-fee ones. As such, I don't expect that other implementations are interpreting the BIP wording as excluding descendants. Good shout, though, since that would be a pretty juicy bug! :bug:

wrt wording used in these comments, I have been trying to use "direct conflicts" as the mempool transactions that a transaction conflicts with, and "original transactions" as all of the mempool transactions that would need to be evicted, i.e. direct conflicts and their descendants. I hope that's clear enough - welcome any feedback on that

ariard replied:
> Which is plural, so i think the implementation is in line with this rule.

Yes "original transactions" is plural though Rule 5 explicitly dissociates "original transactions to be replaced and their descendant transactions". Like the directly-replaced transactions _and_ their descendants as two different things :) ?

I still think this is confusing and our implementation is _not_ in line with this rule. Though I think this is the best behavior for the reason underscored by Gloria.

I'll let someone else checking other full-node implementations if they have bugs around that!

ariard commented on src/policy/rbf.cpp:
```diff
@@ -47,33 +47,127 @@ RBFTransactionState IsRBFOptInEmptyMempool(const CTransaction& tx)
+        // Ideally we'd keep track of the ancestor feerates and make the decision based on that, but
+        // for now requiring all new inputs to be confirmed works.
+        //
+        // Note that if you relax this to make RBF a little more useful, this may break the
```
In case of code relocation like this one, I think it could ease understanding if a reference to `PreChecks` is added, where the calls to `CalculateMempoolAncestors` happen.

ariard reviewed (commented):
Post-merge Code Review ACK 32748da0

darosior reviewed (commented):
late-to-the-party ACK 32748da0f47f7aa9fba78dfb29aa426b14f15624
